src/transforms/transaction.rs

```diff
@@ -0,0 +1,620 @@
+            },
+            _ => {
+                return Err(format!(
+                    "expected numeric value, found: '{}'",
+                    v.to_string_lossy()
+                ));
+            }
+        }
+        Ok(())
+    }
+
+    fn insert_into(self: Box<Self>, k: Atom, v: &mut LogEvent) -> Result<(), String> {
+        Ok(match self.v {
+            NumberMergerValue::Float(f) => v.insert(k, Value::Float(f)),
+            NumberMergerValue::Int(i) => v.insert(k, Value::Integer(i)),
+        })
+    }
+}
+
+//------------------------------------------------------------------------------
+
+trait TransactionValueMerger: std::fmt::Debug + Send + Sync {
+    fn add(&mut self, v: Value) -> Result<(), String>;
+    fn insert_into(self: Box<Self>, k: Atom, v: &mut LogEvent) -> Result<(), String>;
+}
+
+impl From<Value> for Box<dyn TransactionValueMerger> {
+    fn from(v: Value) -> Self {
+        match v {
+            Value::Integer(i) => Box::new(AddNumbersMerger::new(i.into())),
+            Value::Float(f) => Box::new(AddNumbersMerger::new(f.into())),
+            Value::Timestamp(ts) => Box::new(TimestampWindowMerger::new(ts)),
+            Value::Map(_) => Box::new(DiscardMerger::new(v)),
+            Value::Null => Box::new(DiscardMerger::new(v)),
+            Value::Boolean(_) => Box::new(DiscardMerger::new(v)),
+            Value::Bytes(_) => Box::new(DiscardMerger::new(v)),
+            Value::Array(_) => Box::new(DiscardMerger::new(v)),
+        }
+    }
+}
+
+fn get_value_merger(
+    v: Value,
+    m: &MergeStrategy,
+) -> Result<Box<dyn TransactionValueMerger>, String> {
+    match m {
+        MergeStrategy::Sum => match v {
+            Value::Integer(i) => Ok(Box::new(AddNumbersMerger::new(i.into()))),
+            Value::Float(f) => Ok(Box::new(AddNumbersMerger::new(f.into()))),
+            _ => Err(format!(
+                "expected number value, found: '{}'",
+                v.to_string_lossy()
+            )),
+        },
+        MergeStrategy::Append => match v {
+            Value::Bytes(b) => Ok(Box::new(AppendMerger::new(b))),
+            _ => Err(format!(
+                "expected string value, found: '{}'",
+                v.to_string_lossy()
+            )),
+        },
+        MergeStrategy::Array => match v {
+            Value::Array(a) => Ok(Box::new(ArrayMerger::new(a))),
+            _ => Ok(Box::new(ArrayMerger::new(vec![v]))),
+        },
+        MergeStrategy::Discard => Ok(Box::new(DiscardMerger::new(v))),
+    }
+}
+
+//------------------------------------------------------------------------------
+
+struct TransactionState {
+    fields: HashMap<Atom, Box<dyn TransactionValueMerger>>,
+    stale_since: Instant,
+}
+
+impl TransactionState {
+    fn new(e: LogEvent, strategies: &IndexMap<Atom, MergeStrategy>) -> Self {
+        Self {
+            stale_since: Instant::now(),
+            // TODO: all_fields alternative that consumes
```

Comment:
Being forced to clone below which can be removed if we add a consuming version of `all_fields`.